### PR TITLE
Enable parallelism in Crossgen2 and add flags for it in SuperIlc.

### PR DESCRIPTION
**Pull request by SrivastavaAnubhav**
- Add a `parallelism` flag in Crossgen2
- Add a `crossgen2-parallelism` flag in SuperIlc
- Add a low default amount of parallelism for SuperIlc when compiling with Crossgen2

## Perf improvements
These are the results of compiling System.Private.CoreLib.dll with SuperIlc. The SuperIlc build was a debug build, and the crossgen2 build it used was a release build.
```
ReadyToRun.SuperIlc.exe compile-directory -cr C:\runtime\artifacts\tests\coreclr\Windows_NT.x64.Release\Tests\Core_Root -in C:\runtime\artifacts\tests\coreclr\Windows_NT.x64.Release\Tests\Core_Root --nojit --noexe --measure-perf --large-bubble --release -input-file System.Private.Core* -dop 1 --crossgen2-parallelism {1 or 8}
```

### With 8 threads (5 run average)
Total average compilation time: **9314.74 ms**
Phase breakdown (average):
- Loading time: 687.77
- Graph processing time: 3859.30 ms
  - Added 206711 nodes to mark stack
  - Dependency analysis time: 606.09 ms
  - Wall clock JIT time: 2706.99 ms
  - Total JIT time: 24671.84 ms (sum of all threads)
  - 22680 methods JITed
- Emitting time: 4765.06 ms

### With 1 thread (5 run average)
Total average compilation time: **13954.71 ms**
Phase breakdown (average):
- Loading time: 686.73 ms
- Graph processing time: 8871.62 ms
  - Added 206711 nodes to mark
  - Dependency analysis time: 599.27 ms
  - Wall clock JIT time: 7754.44 ms
  - Total JIT time: 7670.19 ms (sum of all threads)
  - 22680 methods JITed
- Emitting time: 4393.62 ms

### CODE REVIEW

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs [hunk 1]
```diff
@@ -264,8 +276,13 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
...
             using (PerfEventSource.StartStopEvents.JitEvents())
             {
-                ConditionalWeakTable<Thread, CorInfoImpl> cwt = new ConditionalWeakTable<Thread, CorInfoImpl>();
-                foreach (DependencyNodeCore<NodeFactory> dependency in obj)
+                ParallelOptions options = new ParallelOptions
+                {
+                    MaxDegreeOfParallelism = _parallelism
+                };
+                Console.WriteLine("Parallelism");
+                Console.WriteLine(_parallelism);
```
Comment (fadimounir) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: I'm guessing these WriteLines were for debugging purposes? :)

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs [hunk 2]
```diff
@@ -264,8 +276,13 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
         {
             using (PerfEventSource.StartStopEvents.JitEvents())
             {
-                ConditionalWeakTable<Thread, CorInfoImpl> cwt = new ConditionalWeakTable<Thread, CorInfoImpl>();
-                foreach (DependencyNodeCore<NodeFactory> dependency in obj)
+                ParallelOptions options = new ParallelOptions
+                {
+                    MaxDegreeOfParallelism = _parallelism
+                };
+                Console.WriteLine("Parallelism");
```
Comment (SrivastavaAnubhav) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: These were for debugging and should be removed, sorry.

File: src/coreclr/src/tools/crossgen2/crossgen2/CommandLineOptions.cs [hunk 3]
```diff
@@ -140,6 +142,11 @@ public static Command RootCommand()
                     // We don't need to override arity here as 255 is the maximum number of generic arguments
                     Argument = new Argument<string[]>()
                 },
+                new Option(new[] { "--parallelism" }, "Number of threads to run")
+                { 
+                    // -1 sets no restriction on processor count
+                    Argument = new Argument<int>(() => -1)
```
Comment (SrivastavaAnubhav) [inline src/coreclr/src/tools/crossgen2/crossgen2/CommandLineOptions.cs]: Would it be better to explicitly set this to Environment.ProcessorCount here?

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs [hunk 4]
```diff
@@ -189,13 +190,21 @@ public sealed class ReadyToRunCodegenCompilation : Compilation
         /// </summary>
         private readonly JitConfigProvider _jitConfigProvider;
 
+        /// <summary>
+        /// We only need one CorInfoImpl per thread, and we don't want to unnecessarily construct them
+        /// because their construction takes a significant amount of time.
+        /// </summary>
+        private readonly ConditionalWeakTable<Thread, CorInfoImpl> corInfoImpls;
```
Comment (fadimounir) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: Nit: Use underscore for field names

Comment (fadimounir) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: What's the difference between having corInfoImpls as a field vs having it as a local like it was before?

File: src/coreclr/src/tools/crossgen2/crossgen2/CommandLineOptions.cs [hunk 5]
```diff
@@ -140,6 +142,11 @@ public static Command RootCommand()
                     // We don't need to override arity here as 255 is the maximum number of generic arguments
                     Argument = new Argument<string[]>()
                 },
+                new Option(new[] { "--parallelism" }, "Number of threads to run")
```
Comment (fadimounir) [inline src/coreclr/src/tools/crossgen2/crossgen2/CommandLineOptions.cs]:
```suggestion
                new Option(new[] { "--parallelism" }, "Maximum number of threads to use during compilation")
```

**Comment by fadimounir** [comment]
LGTM modulo some nit comments

**Comment by fadimounir** [comment]
cc @dotnet/crossgen-contrib

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs (hunk 4 above)
Comment (SrivastavaAnubhav) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: Having it as a local meant that for the next batch of methods we would clean up all the CorInfoImpls after the CWT went out of scope, which is unnecessary.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs (hunk 1 above)
Comment (SrivastavaAnubhav) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: Yep

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs (hunk 4 above)
Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: Thanks

**Review by fadimounir — APPROVED** [review]

File: src/coreclr/src/tools/crossgen2/crossgen2/Program.cs [hunk 6]
```diff
@@ -313,7 +313,8 @@ private int Run()
                     }
                     CompilationBuilder builder = new ReadyToRunCodegenCompilationBuilder(typeSystemContext, compilationGroup, inputFilePath,
                         ibcTuning: _commandLineOptions.Tuning,
-                        resilient: _commandLineOptions.Resilient);
+                        resilient: _commandLineOptions.Resilient,
+                        parallelism: _commandLineOptions.Parallelism);
```
Comment (davidwrighton) [inline src/coreclr/src/tools/crossgen2/crossgen2/Program.cs]: I'd prefer to see this passed to the builder via the fluent interface pattern like the use of UseBackendOptions below.

**Review by davidwrighton — APPROVED** [review]
Approved modulo the nit around fluent api style.

File: src/coreclr/src/tools/crossgen2/crossgen2/Program.cs (hunk 6 above)
Comment (SrivastavaAnubhav) [reply src/coreclr/src/tools/crossgen2/crossgen2/Program.cs]: Sure. Is there a reason that ibcTuning and resilient shouldn't be moved as well? I see that input file path is used to get the input module in the constructor but those aren't.

Comment (davidwrighton) [reply src/coreclr/src/tools/crossgen2/crossgen2/Program.cs]: No, they really should be moved to.

Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/crossgen2/Program.cs]: Thanks for making this change @SrivastavaAnubhav

**Pull request merged**